1169 sync engine block cache

Oscar-Pepper opened the pull request.

nuttycom committed:
Add stub for a `zcash_protocol` crate.

nuttycom committed:
Move `zcash_primitives::{consensus, constants}` to the `zcash_protocol` crate

nuttycom committed:
Move `Amount` and `NonNegativeAmount` types to the `zcash_protocol` crate.

nuttycom committed:
zcash_protocol: Rename value amounts to `Zatoshis` and `ZatBalance`

nuttycom committed:
zcash_protocol: Modify `Zatoshis` to directly wrap a u64

nuttycom committed:
zcash_primitives: Move the `local-consensus` module to the `zcash_protocol` crate.

nuttycom committed:
Fix broken intra-doc links after `zcash_protocol` crate extraction.

nuttycom committed:
zcash_client_backend: Move the `ShieldedProtocol` and `PoolType` types to `zcash_protocol`

nuttycom committed:
zcash_address: Use `zcash_protocol::consensus::NetworkType`

This inverts the dependency relationship between `zcash_protocol` and
`zcash_address`, permitting the network constants (primarily the HRPs)
defined in `zcash_protocol` to be used directly in `zcash_address`
instead of being duplicated.

nuttycom committed:
Expose address generation errors when constructing default addresses

nuttycom committed:
zcash_primitives: Move the `memo` module to the `zcash_protocol` crate.

str4d committed:
zcash_client_backend: Implement async wallet synchronization function

This implements the necessary state machine for taking a wallet in some
arbitrary synchronization status, and fully scanning (the remainder of)
the chain.

Closes zcash/librustzcash#1169.

nuttycom committed:
`zcash_address`: Add support for ZIP 320, TEX addresses.

Oscar-Pepper committed:
zcash_client_backend: Add block cache trait to sync implementation and implement for `MockBlockSource`

zcash_client_sqlite: Implement block cache for `FsBlockDb` and `TestBlockCache`

Oscar-Pepper commented:
- Implements necessary `BlockCache` trait methods for `FsBlockDb` and `TestBlockCache` (formerly `BlockCache` in testing.rs)
- Updates sync.rs to use `BlockCache` trait methods instead of specific `FsBlockDb` methods, now building as cyclic dependencies have been solved
- Necessary changes to get tests to build and pass, changing many to async to accommodate for async `BlockCache` trait methods
- Changes to `scan_cached_blocks` to allow to read from the block cache

fluidvanadium reviewed (commented):
i have triple-checked that this traitification interacts with BlockCache the same way the original PR does
except for deletion which is not yet implemented here

nuttycom committed:
zcash_client_backend: Add `WalletRead::validate_seed`

This adds a mechanism that allows a caller to verify that a given seed
generates the viewing key that is stored in the wallet for a specified
account.

Fixes #1189

nuttycom committed:
zcash_keys: Fix a few problems with no-flags compilation.

nuttycom committed:
Apply documentation suggestions from code review.

Co-authored-by: str4d <thestr4d@gmail.com>

str4d committed:
Merge pull request #1213 from nuttycom/wallet/key_validation

zcash_client_backend: Add `WalletRead::validate_seed`

nuttycom committed:
Merge pull request #1214 from nuttycom/fix/zcash_keys_flagging

zcash_keys: Fix a few problems with no-flags compilation.

nuttycom committed:
Release zcash_keys version 0.1.1

str4d committed:
Merge pull request #1224 from nuttycom/release/zcash_keys-0.1.1

Release zcash_keys version 0.1.1

nuttycom committed:
Apply suggestions from code review

Co-authored-by: str4d <thestr4d@gmail.com>
Co-authored-by: Daira-Emma Hopwood <daira@jacaranda.org>

Oscar-Pepper committed:
zcash_client_backend: Updated CHANGELOG.md

nuttycom committed:
Remove `network_type` calls that are obviated by the blanket impl.

nuttycom committed:
zcash_protocol: Use `BalanceError` instead of `()` for monetary range violations.

nuttycom committed:
Add missing CHANGELOG entries.

nuttycom committed:
Apply suggestions from code review

Co-authored-by: Daira-Emma Hopwood <daira@jacaranda.org>

nuttycom committed:
Add failing test for incorrect zip321 amount parsing.

nuttycom committed:
Fix incorrect zip321 amount parsing.

str4d committed:
Merge pull request #1142 from nuttycom/crate_zcash_protocol

Factor out protocol constants and value types into a `zcash_protocol` crate.

str4d committed:
zcash_protocol 0.1.0

str4d committed:
zcash_address 0.3.2

nuttycom committed:
Merge pull request #1226 from zcash/tex-addrs

zcash_address 0.3.2: ZIP 320 TEX address support

nuttycom committed:
zcash_client_sqlite: Use errors instead of panics for unimplemented or unsupported down migrations.

nuttycom committed:
zcash_client_sqlite: Add backend impl for the Orchard note commitment tree

nuttycom committed:
zcash_client_sqlite: Extract general helpers from Sapling tree logic

Oscar-Pepper committed:
zcash_client_backend: Resolve merge conflicts in CHANGELOG.md and scanning.rs

Oscar-Pepper committed:
zcash_client_backend: Fixed cargo doc build failure when transparent-inputs feature is not specified

(cherry picked from commit ab701118f38ae3471fd9e72ac752907828f3ecf4)

Oscar-Pepper committed:
zcash_client_backend: Fixed cargo doc build failure when transparent-inputs feature is not specified

nuttycom committed:
zcash_client_sqlite: Add `orchard_received_notes` table and update related views.

nuttycom committed:
zcash_client_sqlite: Implement `wallet::orchard::put_received_note`

nuttycom committed:
zcash_client_sqlite: Implement `WalletDb::get_orchard_nullifiers`

nuttycom committed:
zcash_client_sqlite: Deduplicate `WalletDb::get_sapling_nullifiers`

nuttycom committed:
zcash_client_backend: Add Orchard frontier to `AccountBirthday`

nuttycom committed:
zcash_client_sqlite: Support Orchard scanning

str4d committed:
zcash_client_sqlite: Make internal testing functions module-private

This ensures we only access them through `TestState` in future.

nuttycom committed:
Merge pull request #1232 from nuttycom/wallet/orchard_account_birthday

zcash_client_backend: Add Orchard frontier to `AccountBirthday`

str4d committed:
De-duplicate compact Sapling output creation in `TestState`

str4d committed:
Use correct change address in `TestState::generate_next_block_spending`

str4d committed:
Generalise `TestState::generate_*` methods over a `TestFvk` trait

nuttycom committed:
zcash_client_backend: Use `NonNegativeAmount` for `select_spendable_notes`

nuttycom committed:
zcash_client_backend: Generalize `DecryptedOutput` to support Orchard

nuttycom committed:
zcash_protocol: Add `TryFrom<&MemoBytes>` for `Memo`

str4d committed:
`impl TestFvk for orchard::keys::FullViewingKey`

str4d committed:
Avoid excessive Orchard dummies in `TestState::generate_next_block_spending`

nuttycom committed:
zcash_client_sqlite: Generalize `ReceivedNoteId`

nuttycom committed:
zcash_client_sqlite: Add Orchard note commitments in `put_blocks` and `put_block`

AArnott committed:
zcash_client_sqlite: Rework accounts to support imported viewing keys

str4d committed:
Extend `TestState::generate_next_block_spending` with Orchard support

Oscar-Pepper committed:
zcash_client_backend: Added a sync feature to BlockCache for optional tokio deps and moved tokio rt-multi-thread feature to dev-dependencies

nuttycom committed:
Merge pull request #1233 from zcash/test-harness-orchard-fvk

zcash_client_sqlite: Add Orchard support to test harness

str4d committed:
bugfix: Encode transparent IVK in UIVKs instead of transparent FVK

str4d committed:
Add missing feature flag conditional

str4d committed:
Encode UIVK Sapling component with `sapling-crypto`

str4d committed:
CI: Use faster Ubuntu and Windows runners for testing

str4d committed:
CI: Only run CI workflow once per PR

The previous configuration would run CI twice for every PR, on:
- The merge commit with the base branch.
- The tip commit of the PR branch.

The latter is only useful for debugging. However, we still want to run
CI directly on the `main` branch, to ensure that any PRs with runtime
conflicts but not code conflicts get tested in combination.

nuttycom committed:
Merge pull request #1235 from zcash/diverse_accounts

zcash_client_sqlite: Rework accounts to support imported viewing keys

nuttycom committed:
Merge pull request #1237 from zcash/faster-ci

Faster CI

nuttycom committed:
Merge pull request #1234 from nuttycom/wallet/spendable_value

zcash_client_backend: Update `DecryptedTransaction` to support Orchard

str4d committed:
CI: Add caching

str4d committed:
CI: Remove timeouts from rustfmt and clippy

str4d committed:
Merge pull request #1240 from zcash/migration-errors

zcash_client_sqlite: Use errors instead of panics in down-migrations

nuttycom committed:
Merge pull request #1239 from zcash/fasterer-ci

CI: Add caching

nuttycom committed:
Merge pull request #1228 from Oscar-Pepper/fix_cargo_doc_build_failure

Fixed cargo doc build failure

str4d committed:
CI: Trigger on pushes to `main` branch

str4d committed:
Merge pull request #1242 from zcash/fix-ci-trigger

CI: Trigger on pushes to `main` branch

nuttycom committed:
Merge pull request #1241 from nuttycom/pre_orchard_generalizations

zcash_client_sqlite: Generalize internal operations to simplify Orchard support

nuttycom committed:
zcash_client_sqlite: Make sapling_received_notes.recipient_key_scope optional.

We will only consider notes spendable when both the UFVK & key scope are available.

str4d committed:
Merge pull request #1243 from nuttycom/allow_missing_receiving_key_scope

zcash_client_sqlite: Only consider notes spendable when both seed fingerprint & key scope are available.

str4d committed:
Remove `orchard` feature flag from docs.rs builds while not public

str4d committed:
zcash_client_backend 0.11.1

str4d committed:
zcash_client_sqlite 0.9.1

str4d committed:
zcash_client_sqlite: Generalise `wallet::sapling` tests

These are all "single shielded pool" tests, and should pass for any
shielded pool implemented in the wallet.

str4d committed:
zcash_client_sqlite: Generalise `wallet::commitment_tree` tests

str4d committed:
zcash_client_sqlite: Generalise `wallet::scanning` tests

nuttycom committed:
Merge pull request #1246 from zcash/zcb-0.11-zcs-0.9-doc-fixes

`zcash_client_backend 0.11.1`, `zcash_client_sqlite 0.9.1` with doc fixes

nuttycom committed:
Merge pull request #1247 from zcash/zcs-pool-testing

zcash_client_sqlite: Generalise Sapling tests

str4d committed:
zcash_client_sqlite: Replace tuple with struct in `TestState`

str4d committed:
zcash_client_sqlite: Track Orchard commitment tree sizes in `TestState`

str4d committed:
zcash_client_backend: Correctly handle empty frontiers in test birthdays

str4d committed:
Merge pull request #1248 from zcash/zcs-test-state-orchard-tree

zcash_client_sqlite: Track Orchard commitment tree sizes in `TestState`

str4d committed:
zcash_client_sqlite: Add `OrchardPoolTester`, run tests with Orchard

str4d committed:
zcash_client_sqlite: Use a valid length for test seeds

nuttycom committed:
zcash_client_sqlite: Use correct offsets for Orchard commitment positions.

str4d committed:
zcash_client_sqlite: Use `LocalNetwork` for testing

This enables us to activate Sapling and NU5 at the same height, to
simplify reuse of test logic between Sapling and Orchard.

As a side-effect, `zcash_extensions` is removed from the workspace
because it enables the `zfuture` feature flag unconditionally, which
breaks workspace-level builds because this causes the `zfuture` feature
flag on `zcash_protocol` to be enabled without the corresponding feature
flag on `zcash_client_sqlite` being enabled. We will fix this by moving
from feature flags to config flags for unstable features.

str4d committed:
zcash_client_sqlite: Use a seeded RNG in `TestState`

This makes individual instances of `TestState` deterministic, which
helps with debugging.

str4d committed:
CI: Continue tests on error if test failures are in an unstable flag

nuttycom committed:
Merge pull request #1250 from zcash/zcs-orchard-testing

zcash_client_sqlite: Run single-shielded-pool tests on Orchard

str4d committed:
CI: Separate test "state" from the feature flags activated by it

str4d committed:
CI: Stop actively testing ZFuture and instead just check it builds

str4d committed:
Replace `unstable-nu6` and `zfuture` feature flags with cfg flags

Neither of these should have been feature flags, as they gate breaking
changes to the Zcash consensus rules (and in some ways are incompatible
with each other), while feature flags should be additive.

nuttycom committed:
Merge pull request #1251 from zcash/ci-rework

CI: Rework the test states

str4d committed:
zcash_client_backend: Add Orchard change output support to proposals

str4d committed:
zcash_client_sqlite: Fix balance APIs in `TestState` to include Orchard

str4d committed:
zcash_client_sqlite: Fix more test seeds to use valid lengths

str4d committed:
zcash_client_sqlite: Fix `test_with_nu5_birthday_offset` tree setup

nuttycom committed:
Merge pull request #1252 from zcash/orchard-test-fixes

Various fixes to Orchard tests

nuttycom committed:
zcash_client_sqlite: Ensure we identify the sending account in z->t transactions.

nuttycom committed:
zcash_client_sqlite: Make scan_cached_blocks_detects_spends_out_of_order a common single-pool test

nuttycom committed:
zcash_client_sqlite: Generalize chain tests.

nuttycom committed:
zcash_client_sqlite: Add a failing test of cross-pool transfer

nuttycom committed:
Merge pull request #1253 from zcash/more-orchard-tests

zcash_client_sqlite: More Orchard tests

str4d committed:
Merge pull request #1254 from zcash/orchard-tables

zcash_client_sqlite: Add database tables for Orchard

str4d committed:
zcash_client_sqlite: Use preference order for z->t account matching

Co-authored-by: Daira-Emma Hopwood <daira@jacaranda.org>

str4d committed:
Merge pull request #1255 from zcash/orchard-scanning

zcash_client_sqlite: Implement Orchard scanning

Oscar-Pepper committed:
requested changes:
 - Added read error documentation
 - Changed name of `cache_tip` method to `get_tip_height`
 - Improved doc comments

Oscar-Pepper committed:
zcash_client_backend: Resolved merge conflicts in CHANGELOG.md

Oscar-Pepper committed:
Fixed build errors rebasing on changes from #1192
